In [1]:
import sympy as sp

In [2]:
sqrt2 = sp.sqrt(2)
k, l = sp.symbols('k l', integer=True, positive=True)
r = sp.sqrt(6*(k**2 + k*l + l**2))
I = sp.I

e1 = sp.Matrix([[0,  1, 0],
                [-1, 0, 0],
                [0,  0, 0]]) / sqrt2

e2 = sp.Matrix([[0, 0, 0],
                [0, 0, 1],
                [0, -1, 0]]) / sqrt2

e3 = sp.Matrix([[0, 0, -1],
                [0, 0,  0],
                [1, 0,  0]]) / sqrt2

e5 = I * sp.Matrix([[0, 1, 0],
                    [1, 0, 0],
                    [0, 0, 0]]) / sqrt2

e6 = I * sp.Matrix([[0, 0, 0],
                    [0, 0, 1],
                    [0, 1, 0]]) / sqrt2

e7 = I * sp.Matrix([[0, 0, 1],
                    [0, 0, 0],
                    [1, 0, 0]]) / sqrt2

e4 = I * sp.Matrix([[2*l + k, 0, 0],
                    [0, -2*k - l, 0],
                    [0, 0, k - l]]) / r

basis = {
    "e1": e1, "e2": e2, "e3": e3,
    "e5": e5, "e6": e6, "e7": e7,
    "e4": e4,
}

def comm(A: sp.Matrix, B: sp.Matrix) -> sp.Matrix:
    return sp.simplify(A * B - B * A)

def print_commutator_table(basis_dict):
    names = list(basis_dict.keys())
    for i, a in enumerate(names):
        for b in names[i+1:]:
            C = comm(basis_dict[a], basis_dict[b])
            print(f"[{a}, {b}] =")
            sp.pprint(C)
            print()


print_commutator_table(basis)


[e1, e2] =
⎡ 0    0  1/2⎤
⎢            ⎥
⎢ 0    0   0 ⎥
⎢            ⎥
⎣-1/2  0   0 ⎦

[e1, e3] =
⎡0   0     0 ⎤
⎢            ⎥
⎢0   0    1/2⎥
⎢            ⎥
⎣0  -1/2   0 ⎦

[e1, e5] =
⎡ⅈ  0   0⎤
⎢        ⎥
⎢0  -ⅈ  0⎥
⎢        ⎥
⎣0  0   0⎦

[e1, e6] =
⎡      ⅈ⎤
⎢0  0  ─⎥
⎢      2⎥
⎢       ⎥
⎢0  0  0⎥
⎢       ⎥
⎢ⅈ      ⎥
⎢─  0  0⎥
⎣2      ⎦

[e1, e7] =
⎡0   0    0 ⎤
⎢           ⎥
⎢        -ⅈ ⎥
⎢0   0   ───⎥
⎢         2 ⎥
⎢           ⎥
⎢   -ⅈ      ⎥
⎢0  ───   0 ⎥
⎣    2      ⎦

[e1, e4] =
⎡                         √3⋅ⅈ⋅(-k - l)       ⎤
⎢         0            ────────────────────  0⎥
⎢                           _______________   ⎥
⎢                          ╱  2          2    ⎥
⎢                      2⋅╲╱  k  + k⋅l + l     ⎥
⎢                                             ⎥
⎢   √3⋅ⅈ⋅(-k - l)                             ⎥
⎢────────────────────           0            0⎥
⎢     _______________                         ⎥
⎢    ╱  2          2                          ⎥
⎢2⋅╲╱  k  + k⋅l + l         

We choose the following basis for spinors:

$$(1,f_1, f_2, f_3, f_1\wedge f_2, f_2 \wedge f_3, f_3 \wedge f_1, f_1 \wedge f_2 \wedge f_3)$$

We denote by $\varepsilon_i$ multiplication by $f_i$ on the left and by $\iota_i$ contraction by the vectro dual to $f_i$.

Their matrices have the following form:

In [3]:


# ε_i = wedge by f_i on the left:  ε_i(α) = f_i ∧ α
eps1 = sp.Matrix([
    [0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 1, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, -1, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 0],
])

eps2 = sp.Matrix([
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, -1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 1, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 1, 0],
])

eps3 = sp.Matrix([
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [1, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, -1, 0, 0, 0, 0, 0],
    [0, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 1, 0, 0, 0],
])

# ι_i = contraction with e_i dual to f_i 
iota1 = sp.Matrix([
    [0, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 1, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, -1, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
])

iota2 = sp.Matrix([
    [0, 0, 1, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, -1, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1],
    [0, 0, 0, 0, 0, 0, 0, 0],
])

iota3 = sp.Matrix([
    [0, 0, 0, 1, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 1, 0],
    [0, 0, 0, 0, 0, -1, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 1],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0],
])



Next block computes matrices corresponding to Clifford multiplication.

In [4]:
c = [0]*7
c[0] = I*(eps1 + iota1)
c[1] = I*(eps2 + iota2)
c[2] = I*(eps3 + iota3)
c[4] = (iota1 - eps1)
c[5] = (iota2 - eps2)
c[6] = (iota3 - eps3)
c[3] = -I*sp.Matrix([
    [1,0,0,0,0,0,0,0],
    [0,-1,0,0,0,0,0,0],
    [0,0,-1,0,0,0,0,0],
    [0,0,0,-1,0,0,0,0],
    [0,0,0,0,1,0,0,0],
    [0,0,0,0,0,1,0,0],
    [0,0,0,0,0,0,1,0],
    [0,0,0,0,0,0,0,-1],
])

#last matrix is the -i(-1)^deg

c1 = c[0]
c2 = c[1]
c3 = c[2]
c4 = c[3]
c5 = c[4]
c6 = c[5]
c7 = c[6]



In [5]:
ad = [0]*7
ad[0] = sp.simplify(-sqrt2*c2*c3 + sqrt2*c6*c7 - 6*(k+l)/r*c4*c5)
ad[1] = sp.simplify(sqrt2*c1*c3 + 6*k/r*c4*c6 - sqrt2*c5*c7)
ad[2] = sp.simplify(-sqrt2*c1*c2 + 6*l/r*c4*c7 + sqrt2*c5*c6)
ad[3] = sp.simplify(6/r*((k+l)*c1*c5 - k*c2*c6 - l*c3*c7))
ad[4] = sp.simplify(sqrt2*c2*c7 - sqrt2*c3*c6 - 6*(k+l)/r*c1*c4)
ad[5] = sp.simplify(-sqrt2*c1*c7 + 6*k/r*c2*c4 + sqrt2*c3*c5)
ad[6] = sp.simplify(sqrt2*c1*c6 - sqrt2*c2*c5 + 6*l/r*c3*c4)

#We don't divide by 4 because then sympy starts using float


In the next step we compute the matrix for

$$B_0 = \sum_i c_i(\tilde{ad}_i + \hat{\tilde{ad}}_i).$$


B_left denotes $\sum_i c_i\tilde{ad}_i$, B_right denotes $\sum_i c_i\hat{\tilde{ad}}_i$

In [6]:
B_left0 = sum((c[i]*ad[i]for i in range(7)), sp.zeros(8, 8))

B_left0 = B_left0.applyfunc(lambda e: sp.simplify(sp.powsimp(sp.expand(e), force=True)))


Id = sp.eye(8)
B_left = sp.simplify(sp.kronecker_product(B_left0, Id)/3)


B = sum(
    (sp.kronecker_product(c[i], ad[i]) for i in range(7)),
    sp.zeros(c[0].rows*ad[0].rows, c[0].cols*ad[0].cols)
)

B_right = B.applyfunc(lambda e: sp.simplify(sp.powsimp(sp.expand(e), force=True)))


B0 = sp.simplify(B_left + B_right)

We decompose the matrix $B_0$ into blocks and compute the eigenvalues of each one of them. 

In [7]:
from collections import deque

def block_components(A):
    n = A.rows
    # Make an undirected adjacency list from nonzeros of A
    adj = [set() for _ in range(n)]
    # Iterate over nonzero entries 
    nonzero_dict = A.as_mutable().as_sparse().todok()
#    for (i, j) in A.todok().keys() if hasattr(A, "todok") else A.as_mutable().as_sparse().todok().keys():
    for (i, j) in nonzero_dict.keys():
        if i != j:
            adj[i].add(j)
            adj[j].add(i)

    seen = [False]*n
    comps = []
    for s in range(n):
        if not seen[s]:
            q = deque([s])
            seen[s] = True
            comp = []
            while q:
                v = q.popleft()
                comp.append(v)
                for w in adj[v]:
                    if not seen[w]:
                        seen[w] = True
                        q.append(w)
            comps.append(sorted(comp))
    return comps

def permute_to_blocks(A, comps):
    perm = [i for comp in comps for i in comp]
    Ap = A.extract(perm, perm)   
    return perm, Ap


In [8]:
comps = block_components(B0)
A = permute_to_blocks(B0,comps)

eigs = []
for i, comp in enumerate(comps):
    block = B0.extract(comp, comp)
    e = block.eigenvals()
    eigs.append(e)
    print(e)


{-8*sqrt(2): 2, 8*sqrt(2): 2, -sqrt(96*k**2 + 192*k*l + 96*l**2)/sqrt(k**2 + k*l + l**2): 1, sqrt(96*k**2 + 192*k*l + 96*l**2)/sqrt(k**2 + k*l + l**2): 1, -4*sqrt(6)*k/sqrt(k**2 + k*l + l**2): 1, 4*sqrt(6)*k/sqrt(k**2 + k*l + l**2): 1, -4*sqrt(6)*l/sqrt(k**2 + k*l + l**2): 1, 4*sqrt(6)*l/sqrt(k**2 + k*l + l**2): 1}
{-sqrt(48*k**2 + 48*l**2)/sqrt(k**2 + k*l + l**2): 1, sqrt(48*k**2 + 48*l**2)/sqrt(k**2 + k*l + l**2): 1, -sqrt(112*k**2 + 160*k*l + 112*l**2)/sqrt(k**2 + k*l + l**2): 2, sqrt(112*k**2 + 160*k*l + 112*l**2)/sqrt(k**2 + k*l + l**2): 2}
{-sqrt(48*k**2 + 96*k*l + 96*l**2)/sqrt(k**2 + k*l + l**2): 1, sqrt(48*k**2 + 96*k*l + 96*l**2)/sqrt(k**2 + k*l + l**2): 1, -sqrt(112*k**2 + 64*k*l + 64*l**2)/sqrt(k**2 + k*l + l**2): 2, sqrt(112*k**2 + 64*k*l + 64*l**2)/sqrt(k**2 + k*l + l**2): 2}
{-sqrt(96*k**2 + 96*k*l + 48*l**2)/sqrt(k**2 + k*l + l**2): 1, sqrt(96*k**2 + 96*k*l + 48*l**2)/sqrt(k**2 + k*l + l**2): 1, -sqrt(64*k**2 + 64*k*l + 112*l**2)/sqrt(k**2 + k*l + l**2): 2, sqrt(64*k**2

In [9]:
print(eigs)

[{-8*sqrt(2): 2, 8*sqrt(2): 2, -sqrt(96*k**2 + 192*k*l + 96*l**2)/sqrt(k**2 + k*l + l**2): 1, sqrt(96*k**2 + 192*k*l + 96*l**2)/sqrt(k**2 + k*l + l**2): 1, -4*sqrt(6)*k/sqrt(k**2 + k*l + l**2): 1, 4*sqrt(6)*k/sqrt(k**2 + k*l + l**2): 1, -4*sqrt(6)*l/sqrt(k**2 + k*l + l**2): 1, 4*sqrt(6)*l/sqrt(k**2 + k*l + l**2): 1}, {-sqrt(48*k**2 + 48*l**2)/sqrt(k**2 + k*l + l**2): 1, sqrt(48*k**2 + 48*l**2)/sqrt(k**2 + k*l + l**2): 1, -sqrt(112*k**2 + 160*k*l + 112*l**2)/sqrt(k**2 + k*l + l**2): 2, sqrt(112*k**2 + 160*k*l + 112*l**2)/sqrt(k**2 + k*l + l**2): 2}, {-sqrt(48*k**2 + 96*k*l + 96*l**2)/sqrt(k**2 + k*l + l**2): 1, sqrt(48*k**2 + 96*k*l + 96*l**2)/sqrt(k**2 + k*l + l**2): 1, -sqrt(112*k**2 + 64*k*l + 64*l**2)/sqrt(k**2 + k*l + l**2): 2, sqrt(112*k**2 + 64*k*l + 64*l**2)/sqrt(k**2 + k*l + l**2): 2}, {-sqrt(96*k**2 + 96*k*l + 48*l**2)/sqrt(k**2 + k*l + l**2): 1, sqrt(96*k**2 + 96*k*l + 48*l**2)/sqrt(k**2 + k*l + l**2): 1, -sqrt(64*k**2 + 64*k*l + 112*l**2)/sqrt(k**2 + k*l + l**2): 2, sqrt(64*

In [12]:
eigenvalues = []
for i in eigs:
    eigenvalues.extend(list(i.keys()))

for i in eigenvalues:
    print("$$", sp.latex(i), "$$")


$$ - 8 \sqrt{2} $$
$$ 8 \sqrt{2} $$
$$ - \frac{\sqrt{96 k^{2} + 192 k l + 96 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{96 k^{2} + 192 k l + 96 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{4 \sqrt{6} k}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{6} k}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{4 \sqrt{6} l}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{6} l}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{48 k^{2} + 48 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{48 k^{2} + 48 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{112 k^{2} + 160 k l + 112 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{112 k^{2} + 160 k l + 112 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{48 k^{2} + 96 k l + 96 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{48 k^{2} + 96 k l + 96 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{112 k^{2} + 64 k l + 64 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{112 k^{2} + 64 k l + 64 l^{2}}}{\sq

We obtain the list of eigenvalues:

$$ - 8 \sqrt{2} $$
$$ 8 \sqrt{2} $$
$$ - \frac{\sqrt{96 k^{2} + 192 k l + 96 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{96 k^{2} + 192 k l + 96 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{4 \sqrt{6} k}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{6} k}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{4 \sqrt{6} l}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{6} l}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{48 k^{2} + 48 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{48 k^{2} + 48 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{112 k^{2} + 160 k l + 112 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{112 k^{2} + 160 k l + 112 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{48 k^{2} + 96 k l + 96 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{48 k^{2} + 96 k l + 96 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{112 k^{2} + 64 k l + 64 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{112 k^{2} + 64 k l + 64 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{96 k^{2} + 96 k l + 48 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{96 k^{2} + 96 k l + 48 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{64 k^{2} + 64 k l + 112 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{64 k^{2} + 64 k l + 112 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{96 k^{2} + 96 k l + 48 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{96 k^{2} + 96 k l + 48 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{64 k^{2} + 64 k l + 112 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{64 k^{2} + 64 k l + 112 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{48 k^{2} + 48 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{48 k^{2} + 48 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{112 k^{2} + 160 k l + 112 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{112 k^{2} + 160 k l + 112 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{48 k^{2} + 96 k l + 96 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{48 k^{2} + 96 k l + 96 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{\sqrt{112 k^{2} + 64 k l + 64 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{\sqrt{112 k^{2} + 64 k l + 64 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{6} \left(- k - l\right)}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{4 \sqrt{3} \sqrt{k^{2} + 2 k l + 2 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{3} \sqrt{k^{2} + 2 k l + 2 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{4 \sqrt{3} \sqrt{2 k^{2} + 2 k l + l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{3} \sqrt{2 k^{2} + 2 k l + l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{6} k}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{4 \sqrt{3} \sqrt{k^{2} + l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{3} \sqrt{k^{2} + l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{4 \sqrt{3} \sqrt{2 k^{2} + 2 k l + l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{3} \sqrt{2 k^{2} + 2 k l + l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{6} l}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{4 \sqrt{3} \sqrt{k^{2} + 2 k l + 2 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{3} \sqrt{k^{2} + 2 k l + 2 l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ - \frac{4 \sqrt{3} \sqrt{k^{2} + l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{3} \sqrt{k^{2} + l^{2}}}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{6} l}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{6} \left(- k - l\right)}{\sqrt{k^{2} + k l + l^{2}}} $$
$$ \frac{4 \sqrt{6} k}{\sqrt{k^{2} + k l + l^{2}}} $$


We can see that maximal eigenvalue is $8\sqrt2$ ($2\sqrt2$ after dividing by 4 that we ignored at the start). 

As a next step we compute eigenvalues of B_left, which is $\tilde{B}$. 

In [13]:
print(sp.latex(B_left.eigenvals()))

\left\{ - 4 \sqrt{2} : 8, \  4 \sqrt{2} : 8, \  \frac{2 \sqrt{6} k}{\sqrt{k^{2} + k l + l^{2}}} : 16, \  \frac{2 \sqrt{6} l}{\sqrt{k^{2} + k l + l^{2}}} : 16, \  \frac{2 \sqrt{6} \left(- k - l\right)}{\sqrt{k^{2} + k l + l^{2}}} : 16\right\}


$$\left\{ - 4 \sqrt{2} : 8, \  4 \sqrt{2} : 8, \  \frac{2 \sqrt{6} k}{\sqrt{k^{2} + k l + l^{2}}} : 16, \  \frac{2 \sqrt{6} l}{\sqrt{k^{2} + k l + l^{2}}} : 16, \  \frac{2 \sqrt{6} \left(- k - l\right)}{\sqrt{k^{2} + k l + l^{2}}} : 16\right\}$$

We see that $$\eta(\tilde{B}) = 16 sign(k) + 16 sign(l) - 16 sign(k+l),$$

Since we fixed $k,l > 0$, $\eta(\tilde{B}) = 16$.

We compute the signs of the eigenvalues of $B^{\frac{1}{2},\frac{1}{2}}$ below. Unfortunately, it seems to be impossible to compute the eigenvalues of some blocks symbolically, but since we only need to compute their signs, we can compute the characteristic polynomials and use Descartes' rule of signs to determie the number of positive and negative eigenvalues. It is also important that all of the matrices are self-adjoint, so all of the eigenvalues are real. 

In [14]:
S = (B_left + B0/2)
S = S.applyfunc(sp.simplify)

In [15]:
comps = block_components(S)
print(comps)
eigs = []
x = sp.symbols('x')
for i, comp in enumerate(comps):
    block = S.extract(comp, comp)
    print(f"\nBlock {i}: size {block.rows}×{block.cols}")
    #sp.pprint(block)
    if i == 0:
        p = block.charpoly(x)
        f_p = sp.factor(p)
        continue
    p = block.charpoly(x)
    coeffs = p.all_coeffs()
    signs = []
    for cf in coeffs:
        signs.append(sp.sign(cf))
    print(signs)
    if i >= 7:
        sp.pprint(block)
        e = block.eigenvals()
        eigs.append(e)
        print(e)

[[0, 7, 13, 22, 28, 35, 41, 50, 56, 63], [1, 8, 15, 38, 52, 57], [2, 16, 23, 37, 44, 58], [3, 24, 31, 46, 53, 59], [4, 10, 17, 32, 39, 60], [5, 19, 26, 40, 47, 61], [6, 11, 25, 48, 55, 62], [9], [12, 33], [14, 49], [18], [20, 34], [21, 42], [27], [29, 43], [30, 51], [36], [45], [54]]

Block 0: size 10×10

Block 1: size 6×6
[1, 1, -1, -1, 1, 1, -1]

Block 2: size 6×6
[1, -1, -1, 1, 1, -1, -1]

Block 3: size 6×6
[1, -1, -1, 1, 1, -1, -1]

Block 4: size 6×6
[1, -1, -1, 1, 1, -1, -1]

Block 5: size 6×6
[1, 1, -1, -1, 1, 1, -1]

Block 6: size 6×6
[1, -1, -1, 1, 1, -1, -1]

Block 7: size 1×1
[1, 1]
⎡  4⋅√6⋅(-k - l)   ⎤
⎢──────────────────⎥
⎢   _______________⎥
⎢  ╱  2          2 ⎥
⎣╲╱  k  + k⋅l + l  ⎦
{4*sqrt(6)*(-k - l)/sqrt(k**2 + k*l + l**2): 1}

Block 8: size 2×2
[1, 1, -1]
⎡ √6⋅(-3⋅k - 4⋅l)          -√6⋅k       ⎤
⎢──────────────────  ──────────────────⎥
⎢   _______________     _______________⎥
⎢  ╱  2          2     ╱  2          2 ⎥
⎢╲╱  k  + k⋅l + l    ╲╱  k  + k⋅l + l  ⎥
⎢           

We can see that blocks of size 6 must have 3 positive roots and 3 negative roots since the signs of their coefficients are [1, 1, -1, -1, 1, 1, -1] or [1, -1, -1, 1, 1, -1, -1]. There are 3 sign changes, so by Descartes' rule of signs the polynomials must have either 3 or 1 positive roots. After substituting $x \mapsto -x$, there are also 3 sign changes, so there are either 3 or 1 negative signs. Since all roots are real, we can deduce that these blocks do not contribute to the eta invariant. 

Looking at the blocks of size two we can use the same argument or look at the eigenvalues directly. (in one of the cases sympy cannot compute the sign and writes $sign(-k^2 + 2kl - l^2)$, which is negative since we assume that $k \neq l$)

Blocks of size one contribute $2sign(k) + 2sign(l) - 2sign(k+l)$, which is just 2 in our case. 

We deal with 10x10 matrix separately by decomposing its characteristic polynomial. 

In [18]:
u = sp.symbols("u")
a = f_p.subs(sp.sqrt(k**2 + k*l + l**2), u)
#substitute \sqrt{k^2 + kl + l^2) for u
a = sp.cancel(a)
a = sp.simplify(a)

In [19]:
print(sp.latex(sp.factor(a)))

\frac{x^{2} \left(x^{2} - 96\right) \left(x^{2} - 48\right) \left(96 \sqrt{6} k^{2} l x + k^{2} u x^{4} - 144 k^{2} u x^{2} + 3072 k^{2} u + 96 \sqrt{6} k l^{2} x + k l u x^{4} - 144 k l u x^{2} + 3072 k l u + l^{2} u x^{4} - 144 l^{2} u x^{2} + 3072 l^{2} u\right)}{u \left(k^{2} + k l + l^{2}\right)}


$$\frac{x^{2} \left(x^{2} - 96\right) \left(x^{2} - 48\right) \left(96 \sqrt{6} k^{2} l x + k^{2} u x^{4} - 144 k^{2} u x^{2} + 3072 k^{2} u + 96 \sqrt{6} k l^{2} x + k l u x^{4} - 144 k l u x^{2} + 3072 k l u + l^{2} u x^{4} - 144 l^{2} u x^{2} + 3072 l^{2} u\right)}{u \left(k^{2} + k l + l^{2}\right)} = $$

$$ x^{2} \left(x^{2} - 96\right) \left(x^{2} - 48\right) \left(x^4 - 144x^2 +96\sqrt{6} \frac{kl(k+l)}{(k^2+kl+l^2)u}x +3072\right).$$

We see that roots are $\{0,0,\pm \sqrt{96}, \pm\sqrt{48}\}$ and 4 roots from the last polynomial which depend on $k,l$ somehow. Again we can use the sign rule to deduce that it has 2 positive roots and 2 negative roots. 